Website: Product Configurator and Bootstrap 4

netchampfaris opened the pull request:
- Banner Image
- Add item meta tags
- Product page redesign

depends on https://github.com/frappe/frappe/pull/6872


![image](https://user-images.githubusercontent.com/9355208/48349020-073b7a00-e6a9-11e8-9213-99ff5dcc290b.png)

netchampfaris committed:
New Homepage based on bootstrap 4

- Banner Image
- Add item meta tags
- Product page redesign

depends on https://github.com/frappe/frappe/pull/6449

nabinhait committed:
Merge branch 'develop' into website-bootstrap-4

rmehta commented on erpnext/templates/generators/item.html:
```diff
@@ -11,133 +14,212 @@
+				</table>
+			{% endif %}
+		</div>
+		<!-- <div class="item-attribute-selectors">
```
remove commented code

rmehta reviewed (dismissed):
add a few basic tests

netchampfaris committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into website-bootstrap-4

netchampfaris committed:
fix: Move erpnext homepage to home from index

netchampfaris committed:
feat: Products list page

- Item Attributes as configurable filters

netchampfaris committed:
fix: Move products page to www

netchampfaris committed:
feat: Website Filter configuration in Products Settings

1. User can configure which link fields to filter with
2. User can configure which attributes to filter with

Pagination also works on products page

netchampfaris committed:
fix: Split item template into component files

- add configure button for template

netchampfaris committed:
feat: Item configure with attribute filters working

netchampfaris committed:
fix: Update variant_of in attributes table

netchampfaris committed:
fix: Refactor item_cache into a class, identify optional attributes

netchampfaris committed:
feat: Find exact match if there are enough selected attributes

netchampfaris committed:
fix: Styling

netchampfaris committed:
fix: Show attributes with all values and valid_values

- Show disabled attribute values in Select

netchampfaris committed:
feat: Refactor Shopping Cart

- Show items in a table
- Show addresses in cards
- Additional notes field

netchampfaris committed:
feat: Request for Quote if enable_checkout is disabled

netchampfaris committed:
fix: Refactor Item Configure into class

netchampfaris committed:
fix: Bootstrap 4 fixes

netchampfaris committed:
fix: Cleanup homepage

netchampfaris committed:
refactor: Use frappe.parse_json

netchampfaris committed:
feat: Show Item Attribute Table based on setting

netchampfaris committed:
fix: Dont cache products page

netchampfaris committed:
fix: Reset filters button

netchampfaris committed:
fix: Disable configure button when loading dialog

netchampfaris committed:
style: Convert indentation to tabs

netchampfaris committed:
fix: Update valid options on attribute selection

- show all attributes in modal
- show Reset button for optional attributes

netchampfaris committed:
fix: Already selected options are valid so dont disable them

- Cache attribute value selection in case user closes the dialog or
is asked to login

netchampfaris committed:
fix: Cart icon navigates to cart page

netchampfaris committed:
fix: Remove console statement

netchampfaris committed:
feat: Item Inquiry button

- Creates a lead in ERPNext

netchampfaris committed:
fix: Dont show Item Price in cart

- if Enable checkout is not set

netchampfaris committed:
fix: Show continue configuration state on button

- Clear values button

netchampfaris committed:
fix: Truncate item description

- Show more details button

netchampfaris committed:
fix(products page): Optimize for mobile view

netchampfaris committed:
fix: class name

netchampfaris committed:
fix: Order page

netchampfaris committed:
fix: Toggle filter styling

netchampfaris committed:
fix: More Details link route

netchampfaris committed:
fix: Hero section and cards styling

netchampfaris committed:
fix: Move search above product list

netchampfaris committed:
fix: Simple image preview

netchampfaris committed:
fix: Remove redundant description label

netchampfaris committed:
fix: Show clear button for each input

- Show Clear Values even after finding a match

netchampfaris committed:
fix: Translate strings

netchampfaris committed:
fix: Validate Filter Fields

netchampfaris committed:
feat: Allow Filter Fields of type Table MultiSelect

netchampfaris committed:
fix: Filter out invalid attribute values from the initial list of values

netchampfaris committed:
feat: Add Website Content field

- A flexible content field which can be used to render arbitrary HTML
content on the Item Page

netchampfaris committed:
fix: Show inline filter if filter values are more than 20

netchampfaris committed:
fix(wip): Range values in configure

netchampfaris committed:
feat: Introduce Hompage Section

- Can render cards which are provided in the child table
- Can choose how many cards to render in one row
- Can also render custom HTML for full flexibility
- Section Order decides the order in which they appear on the homepage

netchampfaris committed:
fix: Prettify template for job opening

netchampfaris committed:
fix: Use lazy image loading

netchampfaris committed:
fix: Introduce Website Meta Tag into Homepage and Item

netchampfaris committed:
fix: Remove Meta Tags table from Homepage and Item

- Add Set Meta Tags button

netchampfaris committed:
fix: Move erpnext web css to scss

scmmishra reviewed (dismissed):
- [x] Only show published items on /products
- [x] Empty state handling for homepage
- [x] Homepage Section is not discoverable
- [x] Handle image sizes better on /products

netchampfaris committed:
fix: Fallback for homepage title

netchampfaris committed:
fix: Common get_products_settings

netchampfaris committed:
fix: Use new image classes

netchampfaris committed:
fix: Broken or non existent images are now handled by CSS

netchampfaris committed:
fix: Add Homepage Section and Products Settings to desktop config

netchampfaris committed:
fix: Default custom theme content for ERPNext

netchampfaris committed:
fix: Hide Seta meta tags button if website is disabled

netchampfaris committed:
fix(Homepage): Hero Image and Slideshow

netchampfaris committed:
fix: Tag line and description

netchampfaris committed:
fix: Website Image styling

netchampfaris committed:
fix: Add default metatags on homepage

netchampfaris committed:
fix: Empty spaces in additional notes textarea

netchampfaris committed:
fix: Convert product item row to cards for better mobile view

netchampfaris committed:
fix: Better form design for Homepage Section

netchampfaris committed:
fix: Homepage Hero Section refactor

- Better form design
- Hero section based on Default, Slideshow and Homepage Section
- Simple homepage load test

netchampfaris committed:
fix: Additional spaces in cart textarea

netchampfaris committed:
fix: Order product page by weightage

netchampfaris committed:
fix: Patch to set default Homepage Hero Section

netchampfaris committed:
fix(cart): Past orders link and open print in new tab

netchampfaris committed:
fix: Products page query

- Build raw query for nested filters
- Next/Prev navigation fix in filtered view

netchampfaris committed:
fix: Item page

- Show Item Name and price when one item found in configure dialog
- Render Item Price and Stock availability using Jinja
- Remove product_page.js

netchampfaris committed:
fix: Product Settings form design

netchampfaris committed:
fix: Remove Item Variant Table

- Configure dialog is better in terms of UX

netchampfaris committed:
fix: Linting

netchampfaris committed:
fix: Invalidate variants cache on Item update

netchampfaris committed:
fix: comma

netchampfaris committed:
tests: Add test for Homepage Section

netchampfaris committed:
fix: Add Homepage Section link in Homepage doctype

netchampfaris committed:
fix: Cleanup item template, remove item_meta.html

netchampfaris committed:
fix: Unset default value for Website Content

netchampfaris committed:
fix: Cleanup unused css

netchampfaris committed:
fix: Move /products to /all-products

netchampfaris committed:
fix: Allow Homepage Section Rename

netchampfaris committed:
fix: Render price if there is one

netchampfaris committed:
fix: URL in filter

netchampfaris committed:
fix: Move product configurator to portal module

netchampfaris committed:
refactor: Range Input

- Append range input value in additional notes

netchampfaris committed:
fix: Item Group template

- Use item_row for item group list
- Deprecate products_as_list setting

netchampfaris committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into website-bootstrap-4

netchampfaris committed:
fix: Add Item fields again

netchampfaris committed:
fix: Filter fields

- Validate filter fields
- Load valid fields in select

netchampfaris committed:
fix: Attribute filter query

netchampfaris committed:
fix: Handle 0 values

netchampfaris committed:
test: Product Configurator

netchampfaris committed:
fix: Handle empty range_values

netchampfaris committed:
style: Missing semicolon and unused variable

netchampfaris committed:
fix: Remove frappe.set_user statements from tests

netchampfaris committed:
fix: Remove duplicate file

netchampfaris committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into website-bootstrap-4

netchampfaris committed:
fix: Tests

netchampfaris commented:
Codacy is picking up someone else's url, it passed in frappe's [codacy](https://app.codacy.com/app/frappe/erpnext/pullRequest?prid=2540357).

jigneshpshah commented:
Error on website for new Item not having image and web_long_description .

#### Environment :
ERPNext: v12.0.4 (version-12)
Frappe Framework: v12.0.4 (version-12) 

#### Step To Reproduce : 

1. Create an Item 
2. Check -> Show in Website
3. Click on "See on Website" link

### Actual Behavior 

System generates below error

Uncaught Server Exception
```
Traceback (most recent call last):
  File "/home/frappe/frappe-bench/apps/frappe/frappe/website/render.py", line 48, in render
    data = render_page_by_language(path)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/website/render.py", line 152, in render_page_by_language
    return render_page(path)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/website/render.py", line 168, in render_page
    return build(path)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/website/render.py", line 175, in build
    return build_page(path)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/website/render.py", line 197, in build_page
    html = frappe.get_template(context.template).render(context)
  File "/home/frappe/frappe-bench/env/lib/python3.6/site-packages/jinja2/asyncsupport.py", line 76, in render
    return original_render(self, *args, **kwargs)
  File "/home/frappe/frappe-bench/env/lib/python3.6/site-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/home/frappe/frappe-bench/env/lib/python3.6/site-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/frappe/frappe-bench/env/lib/python3.6/site-packages/jinja2/_compat.py", line 37, in reraise
    raise value.with_traceback(tb)
  File "/home/frappe/frappe-bench/apps/erpnext/erpnext/./templates/generators/item/item.html", line 1, in top-level template code
    {% extends "templates/web.html" %}
  File "/home/frappe/frappe-bench/apps/frappe/frappe/./templates/web.html", line 1, in top-level template code
    {% extends base_template_path %}
  File "/home/frappe/frappe-bench/apps/frappe/frappe/./templates/base.html", line 68, in top-level template code
    {% block content %}
  File "/home/frappe/frappe-bench/apps/frappe/frappe/./templates/web.html", line 60, in block "content"
    {{ main_content() }}
  File "/home/frappe/frappe-bench/env/lib/python3.6/site-packages/jinja2/sandbox.py", line 438, in call
    return __context.call(__obj, *args, **kwargs)
  File "/home/frappe/frappe-bench/env/lib/python3.6/site-packages/jinja2/runtime.py", line 579, in _invoke
    rv = self._func(*arguments)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/./templates/web.html", line 15, in template
    {% block page_container %}
  File "/home/frappe/frappe-bench/apps/frappe/frappe/./templates/web.html", line 30, in block "page_container"
    {%- block page_content -%}{%- endblock -%}
  File "/home/frappe/frappe-bench/apps/erpnext/erpnext/./templates/generators/item/item.html", line 15, in block "page_content"
    {% include "templates/generators/item/item_details.html" %}
  File "/home/frappe/frappe-bench/apps/erpnext/erpnext/./templates/generators/item/item_details.html", line 12, in top-level template code
    {% if frappe.utils.strip_html(doc.web_long_description) %}
  File "/home/frappe/frappe-bench/env/lib/python3.6/site-packages/jinja2/sandbox.py", line 438, in call
    return __context.call(__obj, *args, **kwargs)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/utils/data.py", line 607, in strip_html
    return _striptags_re.sub("", text)
TypeError: expected string or bytes-like object

```

4. Put some text in web_long_description , while keeping image blank,  it will show below error

![image](https://user-images.githubusercontent.com/4900815/64424988-30aa1400-d0c8-11e9-8d1c-a7d1bfa63366.png)

### Expected Behavior 
For item not having image and web_long_description , system should not generate error and should display the item page normally

Note : For the above scenario, it worked fine in version 11 , without any issue.
ERPNext: v11.1.59 (version-11)
Frappe Framework: v11.1.54 (version-11)
